### ISSUE TITLE
Babel doesn't work on default foundation-zurb-template

### ISSUE THREAD

**Issue by Diablo-D3** [zurb/foundation-sites#8256]
```
[19:04:27] Failed to load external module babel-register
[19:04:27] Requiring external module babel-core/register
```

This is on Debian sid's builds of node and npm, version 4.3.1 and 1.4.21 respectively.

**Comment by gakimball** [comment]
Did Gulp still run afterwards, or did it crash?

**Comment by Diablo-D3** [comment]
I omitted the backtrace (which I didn't notice at first glance, but might be useful: Unknown plugin "transform-es2015-arrow-functions" specified in .babelrc), then after the backtrace, it says "npm ERR! start: 'gulp'" then "npm ERR! Exit Status 1".

So, yeah, I think it tried to run gulp, and didn't make it.

**Comment by Diablo-D3** [comment]
It seems the error is, in fact, being generated by gulp.

$ node_modules/.bin/gulp --version
[21:17:38] Failed to load external module babel-register
[21:17:38] Requiring external module babel-core/register
[21:17:38] CLI version 1.2.1
[21:17:38] Local version 4.0.0-alpha.2

Yeoman has a closed ticket for this: https://github.com/yeoman/generator-webapp/issues/356

The commit that closed it is a single line fix: https://github.com/yeoman/generator-webapp/commit/840dcbfafa581a9c3ce0ff873705fe7dfcd65eb5

I have not tested this fix yet.

**Comment by Diablo-D3** [comment]
This fixes the failed to load external module bug.

This did not fix the problem with being unable to find transform-es2015-arrow-functions.

**Comment by Diablo-D3** [comment]
I cludged together a fix, but I'm pretty sure this is the wrong way to fix it.

I added [these lines](https://github.com/zurb/foundation-sites/blob/d79008091c6f32ffd9bd6ea60d266c4a76771886/package.json#L21-L29) and changed .babelrc to only read '{ "presets": ["es2015"] }', and the whole thing starts up correctly, just like it did in 6.1.x.

One or the other doesn't fix it alone, which makes literally no sense, seeing as the npm package for the es2015 preset deps on everything listed in the original .babelrc.

**Comment by Diablo-D3** [comment]
@gakimball #8261 has a screenshot identical to what I'm seeing.

**Comment by Diablo-D3** [comment]
Note: my changes to .babelrc make gulp actually start running and serving pages, but it produces broken Javascript. So, yeah, this isn't a fix.

**Comment by Diablo-D3** [comment]
Okay, I got this whole thing to actually startup and compile JS properly.

Unmodified `.babelrc` from foundation-zurb-template, and replace the `babel-preset-es2015` dep in the original `package.json` with:

```
    "babel-plugin-transform-es2015-modules-commonjs": "^6.3.13",
    "babel-plugin-transform-es2015-arrow-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoped-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoping": "^6.4.0",
    "babel-plugin-transform-es2015-classes": "^6.3.15",
    "babel-plugin-transform-es2015-destructuring": "^6.3.15",
    "babel-plugin-transform-es2015-parameters": "^6.3.26",
    "babel-plugin-transform-es2015-shorthand-properties": "^6.5.0",
    "babel-plugin-transform-es2015-spread": "^6.4.0",
    "babel-plugin-transform-es2015-template-literals": "^6.3.13",
    "babel-register": "^6.5.2",
```

foundation-site's own `package.json` lacks `babel-register` and `babel-plugin-transform-es2015-modules-commonjs`, but has the rest of these.

**Comment by erutan** [comment]
Similarly, I got the 

```
events.js:85
      throw er; // Unhandled 'error' event
            ^
ReferenceError: Unknown plugin "transform-es2015-arrow-functions" specified
```

when I tried running gulp with the simple `.bablrc` as suggested in the upgrade guide to a gulp3 zurb template site (e.g.):

```
{
  "presets": ["es2015"],
  "compact": true
}
```

I'm currently porting over some of the changes from the gulp4 zurb template into my project, I'll let you know how things go & if the above fix worked for me (if/when I run into the same problem).

**Comment by Diablo-D3** [comment]
@erutan That doesn't work here. Gulp runs, but it produces nonworking Javascript. I tried it before, thinking that was the correct fix.

My immediately previous comment has produced a working environment for me.

**Comment by erutan** [comment]
Yeah, I want the Gulp4 stuff anyways - read through the queue and was going to add in your change as well (internet got flaky last night so I punted). It's also mentioned as a fix https://github.com/zurb/foundation-sites/issues/8255 so I think it's time to update the docs?

**Comment by erutan** [comment]
I'm getting the following (both from my original project and from a freshly generated generator site) when following the steps from https://github.com/zurb/foundation-sites/issues/8256#issuecomment-189725269

```
gulp
[15:38:11] Failed to load external module babel/register
[...]/Github/test/gulpfile.babel.js:3
import plugins  from 'gulp-load-plugins';
^^^^^^
SyntaxError: Unexpected reserved word
```

**Comment by erutan** [comment]
ok, progress. followed these steps https://github.com/yeoman/generator-webapp/issues/356#issuecomment-126231339 and am now to the following:

```
eruMBP:test erutan$ gulp -v
[17:40:14] Requiring external module babel-register
[17:40:14] CLI version 3.9.1
[17:40:14] Local version 4.0.0-alpha.2
eruMBP:test erutan$ gulp
[17:40:17] Requiring external module babel-register
[17:40:19] Using gulpfile ~/Github/test/gulpfile.babel.js
/usr/local/lib/node_modules/gulp/bin/gulp.js:129
    gulpInst.start.apply(gulpInst, toRun);
                  ^
 
TypeError: Cannot read property 'apply' of undefined
    at /usr/local/lib/node_modules/gulp/bin/gulp.js:129:19
    at nextTickCallbackWith0Args (node.js:415:9)
    at process._tickCallback (node.js:344:13)
    at Function.Module.runMain (module.js:443:11)
    at startup (node.js:134:18)
    at node.js:962:3
```

not sure why gulp-cli is at 3.9.1. deleted it from `node_modules` and ran `npm update`

```
eruMBP:test erutan$ npm update
gulp-cli@0.4.0 node_modules/gulp-cli
├── mute-stdout@0.1.0
└── interpret@0.6.6
gulp@4.0.0-alpha.2 node_modules/gulp
└── gulp-cli@1.2.1
eruMBP:test erutan$ gulp -v
[17:45:09] Requiring external module babel-register
[17:45:09] CLI version 3.9.1
[17:45:09] Local version 4.0.0-alpha.2
```

**Comment by erutan** [comment]
Solved the above by https://github.com/gulpjs/gulp-cli/issues/31#issuecomment-136688725

`npm install -g "gulpjs/gulp-cli#4.0"`

For some reason a manual non global install didn't work (nor obviously npm itself). Wheeee...

**Comment by colin-marshall** [comment]
@erutan I posted the same solution in the other issue, looks like you also came to the same conclusion. Nice work. I just hope that having `gulp 4.0` installed globally will not create issues with projects that were made with `gulp` < 4.0.

**Comment by erutan** [comment]
@colinmarshall yeah I was wondering/worrying about the same thing. Not an expert but the CLI is supposed to be backwards compatible (though not future compatible heh) and I assume/hope that older versions of gulp will still run happily.

**Comment by colin-marshall** [comment]
Besides the `gulp 4.0` issue being a problem @erutan, I think the issue @Diablo-D3 is having is related to permissions. See #8261. 

If you install the Foundation Zurb Template through the Foundation CLI you don't see the output while it runs `npm install` so you are not aware of errors. I asked the user @Diablo-D3

**Comment by colin-marshall** [comment]
Besides the `gulp 4.0` issue being a problem @erutan, I think the issue @Diablo-D3 is having is related to permissions. See #8261. 

If you install the Foundation Zurb Template through the Foundation CLI you don't see the output while it runs `npm install` so you are not aware of errors. I asked the user @Diablo-D3 

[!https://cloud.githubusercontent.com/assets/3061129/13378831/71392ef8-de13-11e5-8919-85a54aaa7c0f.JPG]

**Comment by colin-marshall** [comment]
Besides the `gulp 4.0` issue being a problem @erutan, I think the issue @Diablo-D3 is having is related to permissions. See #8261. 

If you create a new project with the Foundation Zurb Template through the Foundation CLI you don't see the output while it runs `npm install` so you are not aware of errors. @Diablo-D3 said he had the exact same error as showin in the screenshot from @phifa in #8261. I asked @phifa to run `npm install` after they had created a new project with the CLI and it exposed an npm permissions error. See the screenshot:

![Screenshot](https://cloud.githubusercontent.com/assets/3061129/13378831/71392ef8-de13-11e5-8919-85a54aaa7c0f.JPG)

Permissions would make sense in @Diablo-D3's case too because it appears that anything that's a dependency of the main dependencies in the project's `package.json` is not working. That's why adding the packages listed below to `package.json` works. It's because they are now top level dependencies in the project and don't run into the permissions issue.

```
    "babel-plugin-transform-es2015-modules-commonjs": "^6.3.13",
    "babel-plugin-transform-es2015-arrow-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoped-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoping": "^6.4.0",
    "babel-plugin-transform-es2015-classes": "^6.3.15",
    "babel-plugin-transform-es2015-destructuring": "^6.3.15",
    "babel-plugin-transform-es2015-parameters": "^6.3.26",
    "babel-plugin-transform-es2015-shorthand-properties": "^6.5.0",
    "babel-plugin-transform-es2015-spread": "^6.4.0",
    "babel-plugin-transform-es2015-template-literals": "^6.3.13",
    "babel-register": "^6.5.2",
```

I would suggest this to @Diablo-D3:
- Try to fix the permissions issue with the solution here: http://stackoverflow.com/questions/30860795/fail-to-install-npm-package-npm-err-errno-4048/31327161#31327161
- Install gulp 4.0 globally with npm install -g "gulpjs/gulp-cli#4.0"
- Create a new project with the Foundation Zurb Template using the Foundation CLI
- Tell us what happens

**Comment by erutan** [comment]
I ran into the same/similar thing, but wasn't using the CLI (I forked the template, so ran a merge, then just clone it locally and do the manual steps). Permissions may be on since I had to nuke npm/node/gulp to get mine to work, but hard to isolate.

I'd probably just err on the side of a slightly more verbose build time and less possible errors down the road. @Diablo-D3 seems to have gotten his setup working without having to do the global gulp-cli install - though weirdly enough I went into another dir (OS X here) and `gulp -v` gave me `3.9.1`. o_O

**Comment by colin-marshall** [comment]
@erutan Can you please elaborate on this:
> I'd probably just err on the side of a slightly more verbose build time and less possible errors down the road.

Thanks!

**Comment by erutan** [comment]
@colin-marshall, It was in reference to using individual plugins vs the plugin preset. I'm not sure how widespread the upgrade struggle will be, but I imagine there will be more people attempting this next week. Then again perhaps I'm just biased since I just went thought it. ^^;

It'd depend on how much of a gain there is in using the preset I guess.

**Comment by colin-marshall** [comment]
Gotcha. I disagree about there being less possible errors down the road. Including the individual plugins in `package.json` would result in less possible errors only for Foundation projects. If the underlying issue is with bad npm permissions, you're going to run into the same types of errors every time you use npm (not just with Foundation projects), resulting in way more errors down the road. Had you not "gone through it" here, you would have eventually "gone through it" with something else that uses npm.

The best solution is always to fix npm permissions when they are the problem, not to use workarounds because you'll just run into other permissions issues down the road.

**Comment by colin-marshall** [comment]
2 things I think need to happen:
- Foundation CLI needs to inform the user of errors during `npm install`
- Documentation on permissions errors, gulp 4 global (if it's really necessary), etc.

**Comment by erutan** [comment]
I wasn't using Foundation CLI, just npm install, and I got warnings about package deps and whatnot but no hard errors on install iirc. Not sure what install method @Diablo-D3 was using. I'd imagine that permissions errors would work at the registry level, not in the subdirectories of a local project (so not entirely sure how being more verbose on dependencies would offset that). I definitely cleared out my `node_modules` dir a few times (then again I don't really grok how NPM actually works).

At the very least it's worth mentioning if you get the typeerror, do the gulp 4 global thing for now. Even though it didn't install it globally on my side, and that's with a freshly reset environment.

**Comment by fndaily** [comment]
#### using the cli

`$foundation new` 
`$cd the_new_project`

opened up package.json - added this chunk

"babel-plugin-transform-es2015-modules-commonjs": "^6.3.13",
"babel-plugin-transform-es2015-arrow-functions": "^6.3.13",
"babel-plugin-transform-es2015-block-scoped-functions": "^6.3.13",
"babel-plugin-transform-es2015-block-scoping": "^6.4.0",
"babel-plugin-transform-es2015-classes": "^6.3.15",
"babel-plugin-transform-es2015-destructuring": "^6.3.15",
"babel-plugin-transform-es2015-parameters": "^6.3.26",
"babel-plugin-transform-es2015-shorthand-properties": "^6.5.0",
"babel-plugin-transform-es2015-spread": "^6.4.0",
"babel-plugin-transform-es2015-template-literals": "^6.3.13",
"babel-register": "^6.5.2",

`$npm install` 
`$foundation watch`

Repeatable on several flavors of osx including current el capitan beta.

**Comment by fndaily** [comment]
####using the github foundation-sites

`$git clone https://github.com/zurb/foundation-sites.git some_project`
`$npm install`
`$npm start` **(failed)**

on el capitan needed:
`$gem install rvm` 
`$gem install sass`

`$npm start`

**Comment by erutan** [comment]
@fndaily did you try using foundation-sites CLI without making the changes to `package.json`? If so that'd be good to know that the plugin-preset is in fact causing problems.

This thread is about using F6 with the gulp workflow from https://github.com/zurb/foundation-zurb-template, no gems necessary.

**Comment by fndaily** [comment]
@erutan yes indeed - tried everything according to defaults and couldn't get it to work either way. Sorry for the typo.

**Comment by fndaily** [comment]
Ok also tried this backwards 

`$gem uninstall rvm`  
`$npm start` still runs great

`$gem uninstall sass`
`$npm start` **(failed)**

no clue why but this is repeatable - reinstalled gem sass and works fine again.

**Comment by erutan** [comment]
Well you'd need SASS to have the task finish, but you already have a version of ruby on OS X by default.

I suppose the thing to do is to just test the build out on a fresh VM or something to confirm it wasn't existing permissions issues on our machines that were gumming things up.

**Comment by colin-marshall** [comment]
@erutan Foundation uses Libsass (via `gulp-sass`/`node-sass`) not Ruby Sass, so uninstalling the Ruby Sass gem should not make a difference. Not sure what's going on there. I'm pretty sure the only use of Ruby Sass still in Foundation is `gulp-scss-lint`, which is only used in the main `foundation-sites` repo (this repo). 

@fndaily in your most recent comment were you running `npm start` from a Foundation template or a clone of this repo?

**Comment by fndaily** [comment]
Sorry for the confusion that was with `foundation-sites`

for the foundation-zurb-template this the only way it works for me:

`$foundation new --framework sites --template curb`

open package.json - add your chunk from above: 

    "babel-plugin-transform-es2015-modules-commonjs": "^6.3.13",
    "babel-plugin-transform-es2015-arrow-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoped-functions": "^6.3.13",
    "babel-plugin-transform-es2015-block-scoping": "^6.4.0",
    "babel-plugin-transform-es2015-classes": "^6.3.15",
    "babel-plugin-transform-es2015-destructuring": "^6.3.15",
    "babel-plugin-transform-es2015-parameters": "^6.3.26",
    "babel-plugin-transform-es2015-shorthand-properties": "^6.5.0",
    "babel-plugin-transform-es2015-spread": "^6.4.0",
    "babel-plugin-transform-es2015-template-literals": "^6.3.13",
    "babel-register": "^6.5.2",

then again run npm
`$npm install`
`$npm start`

**Comment by colin-marshall** [comment]
@fndaily: Then the reason why `gem uninstall sass` was messing things up is because of the Ruby Sass dependency in `gulp-scss-lint` that is used in `foundation-sites`. That's expected so there is no need to mess with that as it is unrelated and not an issue.

After fixing permissions, but before adding that "chunk" to `package.json`, were you getting errors from `npm install` or `gulp`? If so, what were they?

**Comment by fndaily** [comment]
This is the output of a failed `$npm start` on defaults, before package.json modifications:

```
[21:31:34] Failed to load external module babel-register
[21:31:34] Requiring external module babel-core/register
/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:206
          throw new ReferenceError(messages.get("pluginUnknown", plugin, loc, i, dirname));
          ^

ReferenceError: Unknown plugin "transform-es2015-arrow-functions" specified in "/Users/fnd/Sites/dev/.babelrc" at 0, attempted to resolve relative to "/Users/fnd/Sites/dev"
at /Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:206:17
at Array.map (native)
at Function.normalisePlugins (/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:182:20)

at OptionManager.mergeOptions (/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:298:36)
at OptionManager.addConfig (/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:234:10)
at OptionManager.findConfigs (/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:434:16)
at OptionManager.init (/Users/fnd/Sites/dev/node_modules/babel-core/lib/transformation/file/options/option-manager.js:482:12)
at compile (/Users/fnd/Sites/dev/node_modules/babel-core/node_modules/babel-register/lib/node.js:83:45)
at loader (/Users/fnd/Sites/dev/node_modules/babel-core/node_modules/babel-register/lib/node.js:128:14)
at Object.require.extensions.(anonymous function) [as .js] (/Users/fnd/Sites/dev/node_modules/babel-core/node_modules/babel-register/lib/node.js:138:7)

npm ERR! Darwin 15.3.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "start"
npm ERR! node v5.7.0
npm ERR! npm  v3.6.0
npm ERR! code ELIFECYCLE
npm ERR! foundation-zurb-template@1.0.0 start: `gulp`
npm ERR! Exit status 1```

**Comment by Diablo-D3** [comment]
@gakimball colin-marshall You linked to the wrong screenshot. He apparently added a second one of Powershell, I am not using Powershell or Windows. The screenshot I'm talking about is the first one; @fndaily's comment immediately above also lists the same error and traceback I'm seeing without my fixes.

https://github.com/zurb/foundation-sites/issues/8256#issuecomment-189962364 seems to be Windows only related. `npm` install installs files into the local project fine. That I am not having an issue with. It is when `npm start` is ran, it fails during `gulp` execution.

I'm not even sure how I'd implement that correctly, as I don't have any globally installed modules (not including `npm` deps that get installed by `apt-get`), nor a `~/.npmrc`.

Also, for the record I do not, nor ever have had, gulp installed globally, or any other relevant package. I don't even install `bower` globally, I add it as a dep. Adding `node_modules/.bin` to your `$PATH` is magic.

**Comment by Diablo-D3** [comment]
@erutan Adding `babel-register` alone does not fix it. It fixes the `Failed to load external module babel-register` problem, but it still says `Unknown plugin "transform-es2015-arrow-functions"`.

For some reason babel and/or grunt is not traveling down the `node_modules` dep tree. `transform-es2015-arrow-functions` is clearly `node_modules/babel-preset-es2015/node_modules/babel-plugin-transform-es2015-arrow-functions` along with the rest of the transformers.

The "fix" would be trying to get babel (or grunt, or whatever is throwing these errors up) to travel down the tree. However, for now, I'm using my fix in my environment because I need to get a website for a client done ;)

**Comment by foodgy** [comment]
@Diablo-D3 I did exactly as you (suggested)[https://github.com/zurb/foundation-sites/issues/8256#issuecomment-189725269], but still ```Requiring external module babel-register```.

**Comment by erutan** [comment]
@foodgy try some of the solutions outlined here. I'm not sure if it's just people having off-kilter NPM setups, a flaw in the plugin, or some combination of the two. https://github.com/zurb/foundation-sites/issues/8256#issuecomment-189943105

**Comment by Diablo-D3** [comment]
@foodgy Are you sure you included the line that says `"babel-register": "^6.5.2"`? And then reran `npm install`?

**Comment by foodgy** [comment]
@Diablo-D3 Sure. Well, i've solved it :) Just installed everything both globaly and localy.

**Comment by fndaily** [comment]
That's when you end up with something like 800+ node_modules and definitely not the correct way to do things.

It's looking like @Diablo-D3 has it right - npm dependancies aren't cascading properly.

**Comment by erutan** [comment]
I upgraded NPM and it didn't help, but it started working around time I updated to the latest stable node - someone in another thread mentioned it working on one machine and not another, and there were different versions of NPM/Node on both.

There's definitely something squirrelly with the process.

**Comment by gakimball** [comment]
To address a few things that I don't believe have been mentioned:

- `gulp-babel` will read the configuration you have in `.babelrc` if it exists, even if you pass even a configuration object when calling `$.babel()` in your Gulpfile.
- For Gulp to read a Gulpfile as ES6, it needs a `.babelrc`. Therefore, both your JavaScript build process *and* your Gulpfile have to be transpiled *with the same Babel configuration*.
- Babel's `es2015` preset includes a CommonJS plugin, and one of the things it does is rename any top-level use of `this` to `undefined`.
- what-input uses an AMD wrapper that uses a top-level `this`, which Babel was then breaking. The error in the console reads something like `Cannot read property "root" of undefined`.
- You can configure the CommonJS plugin to *not* convert `this` to `undefined` by setting the `allowTopLevelThis` option to `true`.
- *However*, if you load plugins through a preset, you can't configure anything inside of the preset.
- You could *remove* the CommonJS plugin, but then that breaks the Gulpfile because it uses `import` statements.

So this is a tangled web. Thankfully, we're getting close to an out:

- The creator of what-input published a version 2.0 which just strips the AMD wrapper entirely. That means it wouldn't be affected by Babel's CommonJS plugin.
- That means we can go back to using the `es2015` plugin with all the defaults set.
- That means the `.babelrc` and `package.json` are much simpler: two Babel dependencies, one Babel preset.

We've already merged a PR to bump our what-input dependency to 2.0, so we can deploy that with a version 6.2.1.

Sorry for the hassle, friends, and thanks for your patience :)

**Comment by kball** [comment]
I believe this has been resolved. Please reopen if you're still seeing issues.
